### PR TITLE
feat: build Cypress for darwin-arm64

### LINKED ISSUE #19908: Native support for Apple Silicon (M1 / arm64 architecture)
### What would you like?

The native support of Apple Silicon by distributing an arm64 macos binary, without having to use Rosetta 2 at all.

**Solution 1:** by distributing two distinct binaries according to the macos architecture (x64 or arm64)
**Solution 2:** by distributing an [Universal macOS Binary](https://developer.apple.com/documentation/apple-silicon/building-a-universal-macos-binary)

This is not a duplicate of https://github.com/cypress-io/cypress/issues/4478 because I am focusing only on Apple Silicon binaries, not ARM Linux binaries in general.
This is however a duplicate of https://github.com/cypress-io/cypress/issues/18724 which was closed, but since my request and my proposed solutions are a bit different, I prefer to open a new issue.

### Why is this needed?

@baversjo wrote a very good summary in  https://github.com/cypress-io/cypress/issues/4478:

> Running cypress on an ARM-based M1 mac is pretty slow under rosetta 2. Considering apple no longer seem to selling new intel macs as of q4 2021, I think it's a pretty important feature to add?

On top of that, I think the workaround explained here <https://cypress.io/blog/2021/01/20/running-cypress-on-the-apple-m1-silicon-arm-architecture-using-rosetta-2/> provides a very poor developer experience, having to maintain two different Nodejs installs with Rosetta.

### Proposed solutions

My proposed solution relies on the ability of electron-packager, and thus electron-builder to be able to build macos arm64 binaries on an x64 host, therefore not having to deploy custom CircleCI arm runners, as explained here: https://github.com/electron-userland/electron-builder/issues/5689

#### Solution 1: two distinct binaries: x64 and arm64

`electron-builder.json` becomes:

```json
{
  [...]
  "mac": {
    "target": {
      "target": "zip",
      "arch": [
        "x64",
        "arm64"
      ]
    },
    [...]
  },
  [...]
}
```

I tested: instead of creating one zip file (`build/build/Cypress-9.4.0-mac.zip`), it creates two (`build/build/Cypress-9.4.0-mac.zip` and `build/build/Cypress-9.4.0-arm64-mac.zip`). And as you can see, the binaries are built against the correct architecture:
    
```bash
# Executed on an Intel macbook
➜  cypress git:(develop) ✗ yarn binary-build --platform darwin --version $(node ./scripts/get-next-version.js)
# [...]

➜  cypress git:(develop) ✗ lipo -archs build/build/mac/Cypress.app/Contents/MacOS/Cypress
x86_64
➜  cypress git:(develop) ✗ lipo -archs build/build/mac-arm64/Cypress.app/Contents/MacOS/Cypress 
arm64
```

Then, I suppose we just would have to upload the two zip files, and modify the various mechanisms in place that detect the current arch to download the correct file.
At first sight, the only "tricky" part would be having to patch the [arch](https://www.npmjs.com/package/arch) lib used in `cli/lib/tasks/download.js` (because of https://github.com/feross/arch/issues/19)

Disclamer: I was not able to test the full build process and thus to test the final binaries on both platforms because I am working on a Macbook behind a corporate proxy, so this is very difficult to install all the correct dev dependencies. Also, I am not a contributor yet so I have a very limited knowledge of the project architecture, so maybe I am missing something.

#### Solution 2: one universal app

In this case `electron-builder.json` becomes:

```json
{
  [...]
  "mac": {
    "target": {
      "target": "zip",
      "arch": "universal"
    },
    [...]
  },
  [...]
}
```

And electron-builder builds the app twice, for each arch, and should bundle the two binaries into the same app.
Unfortunately, the "merge" phase failed with this error on my side:
    
```bash
  • packaging       platform=darwin arch=universal electron=15.3.4 appOutDir=/var/folders/1v/csfvz7q15w10lvp83zzlp72w0000gp/T/cypress-build/darwin/build/mac-universal
  ⨯ Command failed with a non-zero return code (1):
lipo /private/var/folders/1v/csfvz7q15w10lvp83zzlp72w0000gp/T/electron-universal-YpnsKo/Tmp.app/Contents/Resources/app/node_modules/@cypress/get-windows-proxy/node_modules/registry-js/build/Release/registry.node /private/var/folders/1v/csfvz7q15w10lvp83zzlp72w0000gp/T/cypress-build/darwin/build/mac-universal--arm64/Cypress.app/Contents/Resources/app/node_modules/@cypress/get-windows-proxy/node_modules/registry-js/build/Release/registry.node -create -output /private/var/folders/1v/csfvz7q15w10lvp83zzlp72w0000gp/T/electron-universal-YpnsKo/Tmp.app/Contents/Resources/app/node_modules/@cypress/get-windows-proxy/node_modules/registry-js/build/Release/registry.node

fatal error: /Library/Developer/CommandLineTools/usr/bin/lipo: /private/var/folders/1v/csfvz7q15w10lvp83zzlp72w0000gp/T/electron-universal-YpnsKo/Tmp.app/Contents/Resources/app/node_modules/@cypress/get-windows-proxy/node_modules/registry-js/build/Release/registry.node and /private/var/folders/1v/csfvz7q15w10lvp83zzlp72w0000gp/T/cypress-build/darwin/build/mac-universal--arm64/Cypress.app/Contents/Resources/app/node_modules/@cypress/get-windows-proxy/node_modules/registry-js/build/Release/registry.node have the same architectures (x86_64) and can't be in the same fat output file  failedTask=build stackTrace=Error: Command failed with a non-zero return code (1):
lipo /private/var/folders/1v/csfvz7q15w10lvp83zzlp72w0000gp/T/electron-universal-YpnsKo/Tmp.app/Contents/Resources/app/node_modules/@cypress/get-windows-proxy/node_modules/registry-js/build/Release/registry.node /private/var/folders/1v/csfvz7q15w10lvp83zzlp72w0000gp/T/cypress-build/darwin/build/mac-universal--arm64/Cypress.app/Contents/Resources/app/node_modules/@cypress/get-windows-proxy/node_modules/registry-js/build/Release/registry.node -create -output /private/var/folders/1v/csfvz7q15w10lvp83zzlp72w0000gp/T/electron-universal-YpnsKo/Tmp.app/Contents/Resources/app/node_modules/@cypress/get-windows-proxy/node_modules/registry-js/build/Release/registry.node
```

I didn't have time to debug further.

#### Which solution should we choose?

I am not a Mac developer so I don't know what is the best practice here. However, I was not able to test but if I understand well, the merged binary size should double in solution 2. This is why I think solution 1 is preferable in my opinion.
And since all the arch detection/distribution is already in place in the project, it should not be difficult to implement.

### PR DESCRIPTION
**Pull request by flotwig**
- Closes #19908 

### User facing changelog

- Cypress is now available natively on Apple Silicon/M1.

### Additional details

- Adds build process utilizing the MacStadium runner.
- Updates CLI to detect arm64 arch properly.
- Removes add-install-comment automation that didn't work on Monterey and replaces it with a local octokit implementation.
- [x] Services PR: https://github.com/cypress-io/cypress-services/pull/4585

### How has the user experience changed?

People can now use a native build which should provide superior performance to running via Rosetta.

Benchmarking highlights:

* Running tests in Electron (cypress run ...) is roughly 2.5x faster
* Running tests in Chrome is improved, but not very much - only ~1.1x faster
* Time between cypress open and when the Launchpad is fully rendered has decreased 3.5x - median dropped from 9.73s to 2.8s
* Time between cypress open --testingType --browser electron and when the Project GUI/App view is fully rendered decreased 3.2x - from 14s to 4.3s

Full benchmark data: https://docs.google.com/spreadsheets/d/1SHA2KgTywB6rNgPx8GfdlwoG9JDyWioo4QGo2QPK93M/edit#gid=0

### Steps to test

- Use the latest pre-release comment on this PR and attempt to install via `npm i` inside of an M1 machine with or without Rosetta.

### PR Tasks

- [x] Have tests been added/updated?
- [x] Has the original issue (or this PR, if no issue exists) been tagged with a release in ZenHub? (user-facing changes only)
- [x] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? https://github.com/cypress-io/cypress-documentation/pull/4568
- [x] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

### CODE REVIEW

File: cli/test/lib/tasks/download_spec.js [hunk 1]
```diff
@@ -335,110 +335,115 @@ describe('lib/tasks/download', function () {
     })
   })
 
-  it('errors on too many redirects', function () {
-    nock('https://aws.amazon.com')
-    .get('/some.zip')
-    .reply(200, () => {
-      return fs.createReadStream(examplePath)
-    })
+  it('errors on too many redirects', async function () {
```
Comment (flotwig) [inline cli/test/lib/tasks/download_spec.js:338]: The changes to this test aren't directly due to the changes in the PR, but this test was actually creating an unawaited Promise previously that was interfering with async stuff added in `util.getRealArch`. I fixed the test by re-stubbing the redirects and awaiting the 2nd download.start.

File: scripts/add-install-comment.js [hunk 2]
```diff
@@ -10,7 +10,8 @@ const {
   getCIName,
```
Comment (flotwig) [inline scripts/add-install-comment.js:10]: The changes in this file bypass/fix https://github.com/cypress-io/release-automations/issues/67, which allows us to remove a package from release-automations: https://github.com/cypress-io/release-automations/pull/69

File: system-tests/projects/config-cjs-and-esm/ts-cjs-with-invalid-esm-only-import/yarn.lock [hunk 3]
```diff
@@ -0,0 +1,47 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
Comment (flotwig) [inline system-tests/projects/config-cjs-and-esm/ts-cjs-with-invalid-esm-only-import/yarn.lock:1]: Not related to this PR but somehow this project made it into develop w/o a yarn.lock, so I added one to fix the CI

**Comment by lmiller1990** [comment]
Wow, even on a toy project with just 2 specs (both failing, but that's not the point):

```sh
# m1
yarn cypress run --component --project ~/code/dump/basic-cypress  7.50s user 2.32s system 138% cpu 7.097 total

# x64
yarn cypress run --component --project ~/code/dump/basic-cypress  36.86s user 4.51s system 197% cpu 20.970 total
```

Around 5x speedup!

**Review by lmiller1990 — COMMENTED** [review]
Looking over the code, it's mostly just configuration - basically the kind of thing I was expect.

I tried the pre-release and it's really quick. I posted some numbers in another comment, but I'm seeing about a 5x speedup on a few toy projects!

**Review by BlueWinds — APPROVED** [review]

File: cli/lib/util.js [hunk 4]
```diff
@@ -460,12 +460,8 @@ const util = {
...
 
   getPlatformInfo () {
-    return util.getOsVersionAsync().then((version) => {
-      let osArch = arch()
-
-      if (osArch === 'x86') {
-        osArch = 'ia32'
-      }
+    return util.getOsVersionAsync().then(async (version) => {
+      let osArch = await this.getRealArch()
```
[the remark this answers is not in the document]
Comment (flotwig) [reply cli/lib/util.js]: [6d1187a](https://github.com/cypress-io/cypress/pull/20686/commits/6d1187a4fb6917879fade768b140f832a6929a10)

**Review by ryanthemanuel — APPROVED** [review]

**Comment by astone123** [comment]
Works great on my M1 🎉

**Comment by alavkx** [comment]
image

**Comment by maikdiepenbroek** [comment]
The pre-release works like a charm! thanks!

**Comment by Kocal** [comment]
I've just updated to 10.2.0 and **everything is SO MUCH FASTER**. Before, I had to wait ~15 sec to open cypress, open the browser, listing specs... now it's literally instantaneous!

Thanks @flotwig for your work! :)

**Pull request merged**